Enable debug info in "complete" bytecode binaries under Linux

nojb opened the pull request:
This is a fix for #9369 for the case of Unix-like systems (Windows case is addressed below). It is a rewrite of some work that had started in #10159.

Let's recall the context. Normal bytecode executables are data files that embed debug information in a separate section next to the bytecode. This section is accessed using the same machinery used to access the bytecode itself (`caml_seek_section`, etc). Bytecode executables built using `--output-complete-exe` (let's call them "complete bytecode executables"), on the other hand, are normal native executables that contains the bytecode as initialized data. Because of this, the bytecode, debug information and any other initialized data in it are no longer easily accessible. As a consequence, "complete" bytecode binaries do not currently support debug information at all.

In principle the debug information could be embedded in the same way as the bytecode is (that is, as initialized program data). This approach can be split in two different steps:

1. Embed the debug information as initialized program data; and
2. adapt the runtime C API to allow accessing debug information stored in initialized program data.

The first point is not trivial because debug information can be large and most C compilers will choke on it if included a a literal in the C file. It is possible to sidestep this issue by using inline assembly and the `.incbin` directive to directly embed a binary blob directly in the generated executable (there are a few other ways of achieving this, but most are rather system-specific).
Unfortunately, the MSVC assembler does not support this directive (but see below for the plan in this case). For Unix-like systems we integrate https://github.com/graphitemaster/incbin, a single (public domain) header file that takes care of abstracting the differences between the handling of this directive between different systems/assemblers/etc.

The second point is dealt with by adding a new runtime (internal) function
```
void caml_set_main_debug_info(void *, int32_t);
```
(where the argument is supposed to be a pointer to a marshalled block and its length) that can be used to set up an alternative source of debug information from initialized executable data. This function is then called from the C file generated when building "complete" bytecode binaries.

Finally, the plan is to handle Windows in the same way. But 1. needs to be adapted because the MSVC linker does not have an `.incbin`-like directive. Instead I proposed an extension of `flexlink` with similar functionality (see https://github.com/alainfrisch/flexdll/pull/93) and use that instead (this approach was suggested by @alainfrisch). Of course the Windows support can be added in a later PR **if** there is an agreement on the general approach.

In terms of implementation, the `configure` script is adapted to test if `incbin.h` actually works at configuration time, and exports a variable `Config.supports_incbin` that is then checked from inside `Bytelink` to decide whether or not to try to use it (in practice all Unix-like systems should implement the directive). If the directive is supported, then the header file `incbin.h` is installed together with the rest of the runtime headers.

nojb committed:
Add incbin.h to repository

nojb committed:
Add configuration support for incbin.h

nojb committed:
autogen

nojb committed:
runtime: add caml_set_main_debug_info()

nojb committed:
complete bytecode binaries: use incbin

nojb committed:
Add test

xavierleroy commented on configure.ac:
```diff
@@ -1298,6 +1299,28 @@ CPPFLAGS="$saved_CPPFLAGS"
+ INCBIN(confdata, __FILE__);
+ void test(void)
+ {
+   (void)confdata_data;
+   (void)confdata_size;
```
If you wanted to be really cute, you'd check that `confdata_size` is big enough and than `confdata_data` starts with `#include`, perhaps with some spaces before.  But maybe this is excessively cute.

xavierleroy commented on runtime/backtrace_byt.c:
```diff
@@ -347,6 +347,48 @@ intnat caml_collect_current_callstack(value** ptrace, intnat* plen,
+    v = Field(datav, i);
+    orig = Long_val(Field (v, 0));
+    evl = Field(v, 1);
+    for (l = evl; l != Val_int(0); l = Field(l, 1)) {
+      value ev = Field(l, 0);
+      Field(ev, EV_POS) = Val_long(Long_val(Field(ev, EV_POS)) + orig);
+    }
+    /* Record event list */
+    Store_field(events, i, evl);
```
There's some cut-and-paste here with code from `read_main_debug_info`.  I'm not sure the code needs to be factored out as an inline helper function.  However, I'd suggest to add back the comment
```
      /* Relocate events in event list */
```

xavierleroy commented on tools/incbin.h:
```diff
@@ -0,0 +1,368 @@
+/**
+ * @file incbin.h
+ * @author Dale Weiler
+ * @brief Utility for including binary files
+ *
+ * Facilities for including binary files into the current translation unit and
+ * making use from them externally in other translation units.
+ */
```
What is the license of this file?

nojb replied:
It is in the public domain, cf https://github.com/graphitemaster/incbin/blob/main/UNLICENSE

xavierleroy reviewed (commented):
For the record: I'm not a big fan of the `-output-complete-exe` mode and I'm still wondering whether the same results could not be achieved in a simpler way.

This said, if one buys the `-output-complete-exe` approach, the proposed PR is pretty neat and probably the best one can do.  Some comments and questions below.

nojb commented:
Thanks for the review @xavierleroy ! Apologies for the delay in addressing your comments.

> For the record: I'm not a big fan of the `-output-complete-exe` mode and I'm still wondering whether the same results could not be achieved in a simpler way.

I too wonder about that. And also whether it is worth rushing this change before G Fawkes' day (given that in practice few people seem to be bothered by it) or whether to use this time to wait and see whether an alternative approach emerges.

Personally, I wouldn't mind letting it stew until after the glaciation.

nojb commented:
A few months later, I also feel that a more principled approach to the whole `-output-complete-exe` issue would be better. As this is not urgent, I move to close.
